lib/models/MatrixAction.js

```diff
@@ -26,6 +30,47 @@ function MatrixAction(type, text, htmlText, timestamp) {
     this.htmlText = htmlText;
     this.ts = timestamp || 0;
 }
+
+MatrixAction.prototype.formatMentions = function(nickUserIdMap) {
+    const regexString = "(" +
+        Object.keys(nickUserIdMap).map((value) => escapeStringRegexp(value)).join("|")
+        + ")";
+    const userRegex = new RegExp(regexString, "igm");
+    const matched = new Set(); // lowercased nicknames we have matched already.
+    let match;
+    let i = 0;
+    while ((match = userRegex.exec(this.text)) !== null && i < MAX_MATCHES) {
+        i++;
+        let matchName = match[1];
+        // Deliberately have a minimum length to match on, so we don't match smaller nicks accidentally.
+        if (matchName.length < MIN_LENGTH_TO_MATCH || matched.has(matchName.toLowerCase())) {
+            continue;
+        }
+        let userId = nickUserIdMap[matchName];
+        if (userId === undefined) {
+            // We might need to search case-insensitive.
+            const nicks = Object.keys(nickUserIdMap).filter((nick) =>
+                nick.toLowerCase() === matchName.toLowerCase()
+            );
+            if (nicks.length === 0) {
+                continue;
+            }
+            userId = nickUserIdMap[nicks[0]];
+            matchName = nicks[0];
+        }
+        // If this is not HTML text already, make it so.
+        if (this.htmlText === undefined) {
+            this.htmlText = this.text;
```

Comment:
This is unsafe

Reply:
This is escaped in L115. If we think it's HTML (dangerous) then htmlText will be defined first.

Reply:
The escaping is fairly far away path-wise, although I'm fine with it being commented to say why it's safe :)